CIDEVSTC-275 Resource Management for Data Processing Resources

lukecampbell opened the pull request.

lukecampbell committed:
Adds create_data_processes

MauriceManning committed:
add transform worker and test as well as new create methods for data process and data process definition

lukecampbell committed:
More data processing work

lukecampbell committed:
Merge branch 'CIDEVSTC-43_2_rebase' of https://github.com/MauriceManning/coi-services into dataprocessing

Conflicts:
	ion/services/sa/process/data_process_management_service.py

lukecampbell committed:
Adds some resource management for data processing

lukecampbell commented on ion/processes/bootstrap/ion_loader.py:
```diff
@@ -1867,6 +1869,13 @@ def _load_ParameterFunctions(self, row):
+        dpd.description = 'Parameter Function Definition for %s' % name
+        dpd.data_process_type = DataProcessTypeEnum.PARAMETER_FUNCTION
+
+        data_process_management.create_data_process_definition_new(dpd, func_id, headers=self._get_system_actor_headers())
```
Every ParameterFunction gets a DataProcessDefinition.

lukecampbell commented on ion/services/sa/process/data_process_management_service.py:
```diff
@@ -146,6 +147,28 @@ def create_data_process_definition(self, data_process_definition=None):
+        else:
+            raise BadRequest('function_definition is not a function type')
+
+        return dpd_id
```
DataProcessDefinitions are just "book keeping" and so the only thing we do in CRUD is just manage the associations.

lukecampbell commented on ion/services/sa/process/test/test_int_data_process_management_service.py:
```diff
@@ -82,6 +82,20 @@ def setUp(self):
         self.dataset_management = self.datasetclient
         self.process_dispatcher = ProcessDispatcherServiceClient()
 
+    def test_create_data_process(self):
```
The test name needs to get renamed.

lukecampbell commented on ion/services/sa/process/test/test_int_data_process_management_service.py:
```diff
@@ -82,6 +82,20 @@ def setUp(self):
+
+        objs, _ = self.rrclient.find_objects(dpd_id, PRED.hasParameterFunction, id_only=False)
+        self.assertEquals(len(objs), 1)
+        self.assertIsInstance(objs[0], ParameterFunction)
```
So far only validates that the association is there.

lukecampbell commented on ion/services/sa/product/data_product_management_service.py:
```diff
@@ -107,6 +107,92 @@ def create_data_product_(self, data_product=None):
+    def create_data_processes(self, data_product_id=''):
+        '''
+        For each data process launched also create a dataprocess for each parameter function in the data product
+        '''
```
For every single data product activated (CM instantiated) we need a data process for all the parameter functions in that coverage model. That's roughly 10-64 data processes per coverage model instance.

MauriceManning replied:
true. we can discuss with Michael and Tim if these should be 'hidden' DPs but it seems to break the model.

lukecampbell commented on ion/services/sa/product/data_product_management_service.py:
```diff
@@ -107,6 +107,92 @@ def create_data_product_(self, data_product=None):
+        self.RR2.assign_dataset_to_data_product_with_has_dataset(dataset_id, data_product_id)
+
+    def assign_data_product_to_data_product(self, data_product_id='', parent_data_product_id=''):
+        validate_true(data_product_id, 'A data product id must be specified')
```
All of this extra stuff is a merge error it looks like.

lukecampbell commented on ion/services/sa/product/data_product_management_service.py:
```diff
@@ -327,6 +413,7 @@ def activate_data_product_persistence(self, data_product_id=''):
         self.update_data_product(data_product_obj)
 
         self._publish_persist_event(data_product_id=data_product_id, persist_on = True)
+        self.create_data_processes(data_product_id)
```
Only run the data process stuff when the CM is instantiated through `activate_data_product`

lukecampbell commented:
@MauriceManning ready for discussion (not merge). I still need to cleanup some stuff.

lukecampbell committed:
Fixes test name and removes merge errors

lukecampbell commented on ion/processes/data/transforms/test/test_transform_worker.py:
```diff
@@ -0,0 +1,306 @@
+        rdt = RecordDictionaryTool(stream_definition_id=self.stream_def_id)
+        rdt['conductivity'] = 1
+        rdt['pressure'] = 2
+        rdt['salinity'] = 8
```
I'm really surprised RDT didn't throw an exception here. You shouldn't be allowed to assigned scalars to arrays...
I think that there would be problems here.

``` python
rdt['time']         = [0] # time should always come first
rdt['conductivity'] = [2]
rdt['pressure']     = [2]
rdt['salinity']     = [8]
```

MauriceManning replied:
Noted, will clean up in next rev.

lukecampbell commented on ion/processes/data/transforms/test/test_transform_worker.py:
```diff
@@ -0,0 +1,306 @@
+
+        rdt = RecordDictionaryTool.load_from_granule(msg)
+        sal_val = rdt['salinity']
+        self.assertTrue( sal_val == 3)
```
This is good for now but in future tests `np.testing` has some great utilities and when two numeric values aren't equal it can even give suggestions as to what happened.

``` python
np.testing.assert_array_equal(sal_val, np.array([3]))
```

lukecampbell commented on ion/processes/data/transforms/test/test_transform_worker.py:
```diff
@@ -0,0 +1,306 @@
+            module='ion.processes.data.transforms.transform_worker',
+            cls='TransformWorker',
+            config=config
+        )
```
This is in the email you asked about. In the old days I used `create_process` and `schedule_process` for this. For tests it's preferred to use `spawn_process`.

[create_process_definition](https://github.com/lukecampbell/coi-services/blob/dataprocessing/ion/services/cei/process_dispatcher_service.py#L220-221)
[create_process](https://github.com/lukecampbell/coi-services/blob/dataprocessing/ion/services/cei/process_dispatcher_service.py#L275-276)
[schedule_process](https://github.com/lukecampbell/coi-services/blob/dataprocessing/ion/services/cei/process_dispatcher_service.py#L299-300)

I've forgotten the subtleties of what each do but I know I used a combination of them to launch the process. That stuff goes in data process management though, not in the test.

lukecampbell commented on ion/processes/data/transforms/transform_worker.py:
```diff
@@ -0,0 +1,210 @@
+        for dp_id, details in self._dataprocesses.iteritems():
+            out_stream_route = details.get('out_stream_route', '')
+            out_stream_id = details.get('out_stream_id', '')
+            publisher = StandaloneStreamPublisher(stream_id=out_stream_id, stream_route=out_stream_route )
```
For processes there's a class that leverages the governance aspect called `StreamPublisher` like [here](https://github.com/lukecampbell/coi-services/blob/dataprocessing/ion/core/process/transform.py#L147-148)

MauriceManning replied:
yes, will modify in the next round.

lukecampbell commented on ion/processes/data/transforms/transform_worker.py:
```diff
@@ -0,0 +1,210 @@
+            self._publisher_map[dp_id] = publisher
+
+    #todo: temporary
+    def download_file(self, url):
```
I'll try to get something better soon.

lukecampbell commented on ion/services/sa/process/data_process_management_service.py:
```diff
@@ -146,6 +147,28 @@ def create_data_process_definition(self, data_process_definition=None):
+        function_definition = self.clients.resource_registry.read(function_id)
+
+        if isinstance(function_definition,  ParameterFunction):
+            data_process_definition.data_process_type = DataProcessTypeEnum.PARAMETER_FUNCTION
```
I override whatever the user specified to maintain resource consistency.

lukecampbell commented on ion/services/sa/process/data_process_management_service.py:
```diff
@@ -308,6 +329,52 @@ def create_data_process(self, data_process_definition_id='', in_data_product_ids
+        else:
+            configuration.process.lookup_docs = self._get_lookup_docs(in_data_product_ids, out_data_product_ids)
+        dproc = DataProcess()
+        dproc.name = 'data_process_%s' % self.get_unique_id()
```
I think that the data process name should be related to what it's doing. Like 'ctd_seawater' or something.

MauriceManning replied:
Yes, I have already changed this locally to be the data process def name concatenated with the input stream name

lukecampbell committed:
Updates ion-definitions:

- Object definitions for data processing
- A few new service methods that will replace existing ones

lukecampbell commented:
Here's a test for the autosample stuff.
http://buildbot.oceanobservatories.org:8010/builders/coi_any/builds/155

lukecampbell commented:
@MauriceManning Ready to merge?
